mitmproxy/addons/maplocal.py

```diff
@@ -0,0 +1,110 @@
+import mimetypes
+import re
+import typing
+import urllib
+from pathlib import Path
+
+from mitmproxy import exceptions
+from mitmproxy import ctx
+from mitmproxy import http
+from mitmproxy.addons.modifyheaders import parse_modify_spec, ModifySpec
+
+
+class MapLocal:
+    def __init__(self):
+        self.replacements: typing.List[ModifySpec] = []
+
+    def load(self, loader):
+        loader.add_option(
+            "map_local", typing.Sequence[str], [],
+            """
+            Replacement pattern of the form "[/flow-filter]/regex/file-or-directory", where
+            the separator can be any character. The @ allows to provide a file path that
+            is used to read the replacement string.
+            """
+        )
+
+    def configure(self, updated):
+        if "map_local" in updated:
+            self.replacements = []
+            for option in ctx.options.map_local:
+                try:
+                    spec = parse_modify_spec(option, True, True)
+                except ValueError as e:
+                    raise exceptions.OptionsError(f"Cannot parse map_local option {option}: {e}") from e
+
+                self.replacements.append(spec)
+
+    def construct_candidate_path(self, base_path, path_components, filename):
+        candidate_path = base_path.joinpath("/".join(path_components + [filename]))
+        return str(candidate_path)
+
+    def sanitize_candidate_path(self, candidate_path, base_path):
```

Comment:
Here's some existing work on that problem:

 - I believe this is what Flask/werkzeug is doing: https://github.com/pallets/werkzeug/blob/0d0d1e7263013c4e14750d8fb91984cf1584e0e0/src/werkzeug/security.py#L202
https://github.com/pallets/werkzeug/blob/09a248649069780b4adfb910fb69f6fb304f2ea3/tests/test_security.py#L69-L82
 - And here's aiohttp: https://github.com/aio-libs/aiohttp/blob/2a924cae6c19dbbfde220bbd461fc446764ad156/aiohttp/web_urldispatcher.py#L585

Reply:
Looking closer at this, it might actually be the best approach to just call werkzeug's [safe_join](https://werkzeug.palletsprojects.com/en/1.0.x/utils/?highlight=safe_join#werkzeug.security.safe_join). We could then also already call sanitization in `file_candidates` so that it only returns a list of safe paths. What do you think?

Reply:
Do we consider a symlink in the user-provided base-dir to a target outside of base-dir a problem?
E.g., the user provides `/local` and the symlink /local/mylink -> /etc/. To consider this we _do_ need I/O access to sanitize paths.

Reply:
I think that's fine. If a user explicitly symlinks that, it may as well work. No need to check the filesystem.

Reply:
Yes, I agree. :+1: